crates/holochain_state/src/mutations.rs

```diff
@@ -746,51 +746,30 @@ pub fn set_receipts_complete_redundantly_in_dht_db(
 }
 
 #[cfg(feature = "unstable-warrants")]
-/// Insert a [`Warrant`] into the Action table.
+/// Insert a [`Warrant`] into the Warrant table.
 pub fn insert_warrant(txn: &mut Transaction, warrant: SignedWarrant) -> StateMutationResult<usize> {
     let warrant_type = warrant.get_type();
     let hash = warrant.to_hash();
     let author = &warrant.author;
-
-    // Don't produce a warrant if one, of any kind, already exists
-    let basis = warrant.dht_basis();
-
-    // XXX: this is a terrible misuse of databases. When putting a Warrant in the Action table,
-    //      if it's an InvalidChainOp warrant, we store the action hash in the prev_hash field.
-    let (exists, action_hash) = match &warrant.proof {
-        WarrantProof::ChainIntegrity(ChainIntegrityWarrant::InvalidChainOp { action, .. }) => {
-            let action_hash = Some(action.0.clone());
-            let exists = txn
-                .prepare_cached(
-                    "SELECT 1 FROM Action WHERE type = :type AND base_hash = :base_hash AND prev_hash = :prev_hash",
-                )?
-                .exists(named_params! {
-                    ":type": WarrantType::ChainIntegrityWarrant,                    
-                    ":base_hash": basis,
-                    ":prev_hash": action_hash,
-                })?;
-            (exists, action_hash)
-        }
-        WarrantProof::ChainIntegrity(ChainIntegrityWarrant::ChainFork { .. }) => {
-            let exists = txn
-                .prepare_cached(
-                    "SELECT 1 FROM Action WHERE type = :type AND base_hash = :base_hash AND prev_hash IS NULL",
-                )?
-                .exists(named_params! {
-                    ":type": WarrantType::ChainIntegrityWarrant,
-                    ":base_hash": basis
-                })?;
-            (exists, None)
-        }
-    };
+    let timestamp = warrant.timestamp;
+    let warrantee = warrant.warrantee.clone();
+
+    // Don't produce a warrant if one, of any kind, already exists for the warrantee.
+    // TODO: Why not?
```

Comment:
I think because one warrant is sufficient to block an agent from the network. Every further warrant is just extra data that the network has to store.

Which is true with permanent warrants only, that logic doesn't hold if warrants can be forgiven... So this is actually a very unhelpful thing to have buried here. I'd leave a TODO with some of these details?

Reply:
https://github.com/holochain/holochain/pull/4937/commits/c99a47d1b442902601c6e1de6f20b601f17a4038